File: frontend/src/utils/ramadanPeriods.ts
export type RamadanPeriodId = 'pre-ramadan' | 'ramadan' | 'post-ramadan';

export interface RamadanPeriodWindow {
  from: string;
  to: string;
  primaryMonth: number;
}

const DB_OVERRIDES: Record<number, { start: string; end: string; source?: string }> = {};

function parseIsoDate(dateIso: string): Date {
  const [y, m, d] = dateIso.split('-').map(Number);
  return new Date(y, m - 1, d);
}

function toIsoDate(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

function addDays(dateIso: string, days: number): string {
  const dt = parseIsoDate(dateIso);
  dt.setDate(dt.getDate() + days);
  return toIsoDate(dt);
}

function firstDayOfMonth(dateIso: string): string {
  const dt = parseIsoDate(dateIso);
  return toIsoDate(new Date(dt.getFullYear(), dt.getMonth(), 1));
}

function lastDayOfMonth(dateIso: string): string {
  const dt = parseIsoDate(dateIso);
  return toIsoDate(new Date(dt.getFullYear(), dt.getMonth() + 1, 0));
}

export function getRamadanRange(year: number): { start: string; end: string } | null {
  const dbOverride = DB_OVERRIDES[year];
  if (dbOverride) {
    return { start: dbOverride.start, end: dbOverride.end };
  }
  return null;
}

export function setRamadanDateOverride(
  year: number,
  startDate: string,
  endDate: string,
  source?: string,
): void {
  DB_OVERRIDES[year] = { start: startDate, end: endDate, source };
}

export function clearRamadanDateOverride(year: number): void {
  delete DB_OVERRIDES[year];
}

export function hasRamadanDatesConfigured(year: number | undefined): boolean {
  if (!year) return false;
  return Boolean(DB_OVERRIDES[year]);
}

export function getRamadanPeriodWindows(year: number): Record<RamadanPeriodId, RamadanPeriodWindow> | null {
  const range = getRamadanRange(year);
  if (!range) return null;
  const { start, end } = range;
  const preTo = addDays(start, -1);
  const postFrom = addDays(end, 1);

  return {
    'pre-ramadan': {
      from: firstDayOfMonth(start),
      to: preTo,
      primaryMonth: parseIsoDate(start).getMonth() + 1,
    },
    ramadan: {
      from: start,
      to: end,
      primaryMonth: parseIsoDate(start).getMonth() + 1,
    },
    'post-ramadan': {
      from: postFrom,
      to: lastDayOfMonth(end),
      primaryMonth: parseIsoDate(end).getMonth() + 1,
    },
  };
}

export function getRamadanPeriodWindow(
  year: number | undefined,
  month: number | undefined,
  selectedPeriod: string | null | undefined,
): RamadanPeriodWindow | null {
  if (!year || !month || !selectedPeriod) return null;
  if (selectedPeriod !== 'pre-ramadan' && selectedPeriod !== 'ramadan' && selectedPeriod !== 'post-ramadan') {
    return null;
  }

  const windows = getRamadanPeriodWindows(year);
  if (!windows) return null;
  const window = windows[selectedPeriod];
  const fromMonth = parseIsoDate(window.from).getMonth() + 1;
  const toMonth = parseIsoDate(window.to).getMonth() + 1;
  if (month !== fromMonth && month !== toMonth && month !== window.primaryMonth) return null;
  return window;
}

export function isDateInWindow(dateIso: string, window: RamadanPeriodWindow): boolean {
  const parsed = dateIso.split('T')[0];
  return parsed >= window.from && parsed <= window.to;
}

File: backend/ramadan_periods.py
from datetime import date, timedelta
from typing import Dict, Optional, Tuple
from sqlalchemy.orm import Session

from backend.models import RamadanDate

RamadanPeriodId = str

def _first_day_of_month(dt: date) -> date:
    return dt.replace(day=1)


def _last_day_of_month(dt: date) -> date:
    if dt.month == 12:
        return date(dt.year, 12, 31)
    return date(dt.year, dt.month + 1, 1) - timedelta(days=1)


def _load_override_from_db(year: int, db: Optional[Session]) -> Optional[Tuple[date, date]]:
    if db is None:
        return None
    row = db.query(RamadanDate).filter(RamadanDate.year == int(year)).first()
    if row is None:
        return None
    return row.start_date, row.end_date


def get_ramadan_range(year: int, db: Optional[Session] = None) -> Optional[Tuple[date, date]]:
    return _load_override_from_db(year, db)


def get_ramadan_period_windows(year: int, db: Optional[Session] = None) -> Optional[Dict[RamadanPeriodId, Tuple[date, date]]]:
    rng = get_ramadan_range(year, db)
    if rng is None:
        return None
    start, end = rng
    return {
        "pre-ramadan": (_first_day_of_month(start), start - timedelta(days=1)),
        "ramadan": (start, end),
        "post-ramadan": (end + timedelta(days=1), _last_day_of_month(end)),
    }


def get_ramadan_period_window(
    year: Optional[int], month: Optional[int], selected_period: Optional[str], db: Optional[Session] = None
) -> Optional[Tuple[date, date]]:
    if year is None or month is None or not selected_period:
        return None
    windows = get_ramadan_period_windows(year, db)
    if windows is None:
        return None
    window = windows.get(selected_period)
    if window is None:
        return None
    start_d, end_d = window
    if month not in (start_d.month, end_d.month):
        return None
    return window


def detect_periods_for_dates(year: int, month: int, dates: list[date], db: Optional[Session] = None) -> list[str]:
    if not dates:
        return []
    windows = get_ramadan_period_windows(year, db)
    if windows is None:
        return []
    detected: list[str] = []
    for period in ("pre-ramadan", "ramadan", "post-ramadan"):
        start_d, end_d = windows[period]
        if month not in (start_d.month, end_d.month):
            continue
        if any(start_d <= d <= end_d for d in dates):
            detected.append(period)
    return detected

